File: typescript/ai-sloppy-paste/src/utils/snippet-use.ts
import { Clipboard } from "@raycast/api";
import { PlaceholderValueToRecord, recordSnippetUse } from "./storage";

type BestEffortOperation = () => unknown | Promise<unknown>;
type FailureHandler = (error: unknown) => unknown | Promise<unknown>;

interface SnippetActionOptions<T> {
  prepare: () => T | Promise<T>;
  primaryOperation: (prepared: T) => Promise<void>;
  snippetId: string;
  placeholderValues?: PlaceholderValueToRecord[];
  onPreparationFailure: FailureHandler;
  onPrimaryFailure: FailureHandler;
}

interface CopySnippetContentOptions {
  snippetId: string;
  content: string;
  onPrimaryFailure: FailureHandler;
}

/**
 * Runs an action whose failure must not change the outcome of a successful
 * clipboard operation. Errors are intentionally logged without the original
 * error object because it might contain snippet content or placeholder values.
 */
export async function runBestEffort(operation: BestEffortOperation, failureLogMessage: string): Promise<void> {
  try {
    await operation();
  } catch {
    console.error(failureLogMessage);
  }
}

/**
 * Records metadata after a successful clipboard action without exposing
 * snippet content or placeholder values if storage is unavailable.
 */
export async function recordSnippetUseBestEffort(
  snippetId: string,
  placeholderValues: PlaceholderValueToRecord[] = [],
): Promise<void> {
  await runBestEffort(
    () => recordSnippetUse(snippetId, placeholderValues),
    `Unable to record snippet use: ${snippetId}`,
  );
}

/**
 * Keeps preparation and the single primary clipboard operation as explicit
 * failure boundaries. Tracking runs only after the primary operation succeeds.
 */
export async function runSnippetAction<T>({
  prepare,
  primaryOperation,
  snippetId,
  placeholderValues,
  onPreparationFailure,
  onPrimaryFailure,
}: SnippetActionOptions<T>): Promise<boolean> {
  let prepared: T;
  try {
    prepared = await prepare();
  } catch (error) {
    await runBestEffort(() => onPreparationFailure(error), "Unable to show snippet preparation failure");
    return false;
  }

  try {
    await primaryOperation(prepared);
  } catch (error) {
    await runBestEffort(() => onPrimaryFailure(error), "Unable to show primary clipboard failure");
    return false;
  }

  await recordSnippetUseBestEffort(snippetId, placeholderValues);
  return true;
}

/**
 * Shared policy for secondary views that copy full stored snippet content.
 * Clipboard metadata helpers (titles, analytics summaries, syntax examples,
 * and form helpers) deliberately do not use this path and remain untracked.
 */
export async function copySnippetContent({
  snippetId,
  content,
  onPrimaryFailure,
}: CopySnippetContentOptions): Promise<boolean> {
  return runSnippetAction({
    prepare: () => content,
    primaryOperation: (snippetContent) => Clipboard.copy(snippetContent),
    snippetId,
    onPreparationFailure: () => undefined,
    onPrimaryFailure,
  });
}

File: swift/sloppy-paste/Sources/SloppyCore/Use/SnippetUse.swift
import Foundation

/// Failure boundaries for a snippet action: preparation, then one primary
/// clipboard operation, then best-effort use tracking. Log messages carry only
/// safe metadata, never snippet content or placeholder values.
public struct SnippetUse: Sendable {
    /// Records one use of a snippet with its submitted placeholder values.
    public var record: @Sendable (_ snippetID: String, _ placeholderValues: [PlaceholderValueToRecord]) async throws -> Void
    public var log: @Sendable (String) -> Void

    public init(
        record: @escaping @Sendable (String, [PlaceholderValueToRecord]) async throws -> Void,
        log: @escaping @Sendable (String) -> Void = { message in
            FileHandle.standardError.write(Data((message + "\n").utf8))
        }
    ) {
        self.record = record
        self.log = log
    }

    /// Runs an operation whose failure must not change an already-successful outcome.
    public func runBestEffort(_ failureLogMessage: String, _ operation: @Sendable () async throws -> Void) async {
        do {
            try await operation()
        } catch {
            log(failureLogMessage)
        }
    }

    public func recordUseBestEffort(snippetID: String, placeholderValues: [PlaceholderValueToRecord] = []) async {
        await runBestEffort("Unable to record snippet use: \(snippetID)") {
            try await record(snippetID, placeholderValues)
        }
    }

    /// Returns true only when preparation and the primary operation both
    /// succeeded; tracking runs only after that.
    public func run<Prepared: Sendable>(
        snippetID: String,
        placeholderValues: [PlaceholderValueToRecord] = [],
        prepare: @Sendable () async throws -> Prepared,
        primaryOperation: @Sendable (Prepared) async throws -> Void,
        onPreparationFailure: @Sendable (any Error) async throws -> Void,
        onPrimaryFailure: @Sendable (any Error) async throws -> Void
    ) async -> Bool {
        let prepared: Prepared
        do {
            prepared = try await prepare()
        } catch {
            await runBestEffort("Unable to show snippet preparation failure") { try await onPreparationFailure(error) }
            return false
        }

        do {
            try await primaryOperation(prepared)
        } catch {
            await runBestEffort("Unable to show primary clipboard failure") { try await onPrimaryFailure(error) }
            return false
        }

        await recordUseBestEffort(snippetID: snippetID, placeholderValues: placeholderValues)
        return true
    }

    /// Copies full stored content (secondary views) and records one use.
    public func copyContent(
        snippetID: String,
        content: String,
        copy: @Sendable (String) async throws -> Void,
        onPrimaryFailure: @Sendable (any Error) async throws -> Void
    ) async -> Bool {
        await run(
            snippetID: snippetID,
            prepare: { content },
            primaryOperation: copy,
            onPreparationFailure: { _ in },
            onPrimaryFailure: onPrimaryFailure
        )
    }
}
